# Coup de Grâce: Level 3 🔥🔥🔥

Oppgavene er delt opp i **to hovedkategorier**:
1. Oppgaver der du skal tenke deg fram til svaret uten å bruke programmering. Markert med 📝.
2. Oppgaver der du skal bruke programmering til å finne svaret. Markert med 💻.
3. Noen oppgaver er en miks av de to, markert med 💽.

## Oppgave 1



En elev jobber med en oppgave:

````{admonition} Oppgaven
:class: note
En funksjon $f$ er gitt ved

$$
f(x) = 3x^3 - \frac{13}{5}x^2 - \frac{25}{5}x + 1.
$$

Grafen til $f$ er vist under.


```{figure} ./figurer/halveringsmetoden_graf.png
---
name: tredjegradsfunksjon
---
```

Bestem nullpunktene til $f$.
````

Eleven har laget en kode for å løse oppgaven:

In [ ]:
def f(x):
    return 3 * x**3 - 13 * x**2 / 5 - 25 * x / 5 + 1


a = 0
b = 1
m = (a + b) / 2

while abs(f(m)) > 1e-8:     # Betingelsen leses som |f(m)| > 10**(-8)
    if f(a) * f(m) < 0:
        b = m
    else:
        a = m

    m = (a + b) / 2

print(f"{m = :.2f}")

### Oppgave 1a 📝
**Svar på følgende spørsmål uten å kjøre programmet:**
1. Forklar strategien som eleven har brukt i programmet.
2. Hvilket nullpunkt er det programmet finner (bruk grafen over til å forklare hvor, du trenger ikke regne ut nullpunktets verdi)?


````{admonition} Mattehint: Hvor ligger gjennomsnittet av to tall i forhold til de to tallene?
:class: tip, dropdown

Det viser seg at gjennomsnittet av to tallet *alltid* ligger mellom to tall. <br>

Klarer du vise at dette stemmer?

```{admonition} Bevis for at gjennomsnittet av to tall ligger midt mellom de to tallene
:class: dropdown

Vi tenker oss et tall $a$ og et tall $b$. Gjennomsnittet $m$ er da

$$
m = \frac{a + b}{2}.
$$

Avstanden fra $a$ til $m$ vil være

$$
|m - a| = \left|\frac{a + b}{2} - a\right| = \left|\frac{a + b - 2a}{2}\right| = \left|\frac{b - a}{2}\right| = \frac{b-a}{2}
$$

Tilsvarende har vi at avstanden mellom $m$ og $b$ er

$$
|m - b| = \left|\frac{a + b}{2} - b\right| = \left|\frac{a + b - 2b}{2}\right| = \left|\frac{a - b}{2}\right| = \frac{b-a}{2}
$$

De to avstandene er de samme, som må bety at $m$ ligger midt mellom $a$ og $b$. Med andre ord er $m$ midtpunktet på intervallet $[a, b]$. 
```
````

````{admonition} Kodehint: Hva er abs(f(m))? 
:class: dropdown, tip

Funksjonen `abs()` regner ut absoluttverdien av argumentet som sendes inn. `abs(f(m))` betyr det samme som absoluttverdien i matematikk $|f(m)|$.

````


````{admonition} Løsningsforslag
:class: dropdown
1. Strategien til eleven går som følger. Gitt et intervall $[a, b]$ der $f(a)$ og $f(b)$ har forskjellig fortegn, så må det være et punkt på intervallet der hvor grafen til $f$ skjærer $x$-aksen. Derfor finnes det et nullpunkt $x \in [a, b]$. 
Tanken bak programmet er at hvis vi regner ut midtpunktet $m = (a + b) / 2$, så vil vi kunne lage oss et nytt intervall som er halvparten så stort. Da vil grafen til $f$ enten skjærer $x$-aksen i intervallet $[a, m]$ eller i intervallet $[m, b]$. Vi velger oss bare det nye intervallet der $f$ har forskjellig fortegn i endepunktene, for da er vi sikre på at grafen til $f$ skjærer $x$-aksen i det nye intervallet. Gjentar vi dette tilstrekkelig mange ganger blir intervallet såpass lite at vi vil til slutt skvise oss så nærme punktet der $f$ skjærer $x$-aksen som vi bare måtte ønske. Med andre ord vil vi til slutt komme så nærmere et nullpunkt til $f$ som vi bare måtte ønske.
2. Programmet starter med $a = 0$ og $b = 1$. Det er bare ett nullpunkt på dette intervallet som betyr at vi vil nærme oss nullpunktet som ligger på intervallet $[0, 1]$ med programmet.
````

### Oppgave 1b  💽
1. Hvilket nullpunkt finner programmet i stedet hvis vi setter $a = 1$ og $b = 2$? 📝
2. Test forutsigelsen din ved å kjøre programmet med $a = 1$ og $b = 2$.  💻


````{admonition} Løsningsforslag
:class: dropdown

Det er bare ett nullpunkt på intervallet $[1, 2]$, som vi kan se fra grafen til $f$ i oppgaven. Derfor vil programmet nærme seg nullpunktet på intervallet $[1, 2]$. 
Kjører vi programmet med `a = 1` og `b = 2` får vi utskriften

```console
m = 1.72
```
som betyr at $x \approx 1.72$ er et nullpunkt for $f$.
````

### Oppgave 1c 💻

Gjør nødvendige endringer i programmet slik at du finner bunnpunktet til $f$. <br>
Finn bunnpunktet med programmet.

```{admonition} Mattehint: Hva vet du om $f'(x)$ i et bunnpunkt? 
:class: dropdown, tip

I et bunnpunkt vil $f'(x) = 0$. 
```


```{admonition} Kodehint: Hvordan kan du regne ut $f'(x)$ i Python?
:class: dropdown, tip

Det kan være lurt å lage seg en Pythonfunksjon `f_derivert(x)` som regner ut $f'(x)$ for deg. 
```



````{admonition} Løsningsforslag 
:class: dropdown

Først bør vi lage en pythonfunksjon for $f'(x)$:
```python
def f_derivert(x):
    return 9 * x**2 - 26 * x / 5 - 25 / 5
```
Deretter må vi endre betingelsene i `if`-`else`-testen slik at vi bruker `f_derivert` og ikke `f` når vi leter etter et nullpunkt. <br>
Videre må gjøre er å velge oss et intervall der hvor bunnpunktet ligger. Vi ser fra grafen at bunnpunktet ligger på intervallet $[0, 2]$. Derfor kan vi sette `a = 0` og `b = 2` i programmet. <br>
Til slutt må vi skrive ut funksjonverdien til $f$ i toppunktet. 
En mulig løsning av programmet blir derfor:

```python
def f(x):
    return 3 * x**3 - 13 * x**2 / 5 - 25*x / 5 + 1

def f_derivert(x):
    return 9 * x**2 - 26 * x / 5 - 25 / 5

a = 0
b = 2
m = (a + b) / 2

while abs(f_derivert(m)) > 1e-8:
    if f_derivert(a) * f_derivert(m) < 0:
        b = m
    else:
        a = m
        
    m = (a + b) / 2


print(f"{m = :.2f}")
print(f"{f(m) = :.2f}")
```
som gir utskriften

```console
m = 1.09
f(m) = -3.65
```
som betyr at bunnpunktet ligger omtrent i $(1.09, -3.65)$.
````

## Oppgave 2

Nok en elev jobber med en oppgave:

````{admonition} En oppgave som ville gjort Arkimedes sjalu 🏺
:class: note

En funksjon $f$ er gitt ved

$$
f(x) = x^3 + x^2 - 4x + 10
$$

Figuren under viser grafen til $f$ i intervallet $[0, 4]$. Området mellom grafen til $f$ og $x$-aksen er fargelagt.

```{figure} ./figurer/areal.png
---
name: areal_graf
---
```

Bestem arealet av det fargelagte området mellom grafen til $f$ og $x$-aksen.
````

Eleven har laget seg en figur som viser hvordan eleven har tenkt å bestemme en tilnærming til arealet ved hjelp av arealene av rektangeler.

```{figure} ./figurer/areal_figurer.png
---
name: areal
---
```

Eleven har startet på en kode som skal regne ut arealet. Koden ser slik ut:

In [ ]:
def f(x):
    return x**3 + x**2 - 4*x + 10

a = 0
b = 4
bredde = (b - a) / 3                    # Bredde med 3 rektangler

areal = 0
for i in range(3):
    høyde = 0
    areal = 0

print(f"{areal = :.2f}")                # Skriver ut arealet med 2 desimaler

In [3]:
def f(x):
    return x**3 + x**2 - 4*x + 10

a = 0
b = 4
n_rektangler = 1_000_000
bredde = (b - a) / n_rektangler                    

areal = 0
for i in range(n_rektangler):
    høyde = f(a + i*bredde)
    areal = areal + bredde * høyde

print(f"{areal = :.2f}")                # Skriver ut arealet med 2 desimaler

areal = 93.33


### Oppgave 2a 📝
**Svar på følgende spørsmål uten å kjøre programmet:**
1. Forklar hvordan strategien til eleven fungerer. 
2. Regn ut arealet du vil få dersom du bruker like mange rektangler som eleven prøvde å bruke i programmet over.


````{admonition} Løsningsforslag
:class: dropdown

1. Eleven har tenkt at man kan tilnærme arealet undere grafen ved å lage rektangler som har bredde $(b - a) / n$, der $n$ er antall rektangler over intervallet $[a, b]$, og høyden til hvert rektangel er gitt ved funksjonsverdien i venstre hjørne på hvert rektangel.
2. Eleven har valgt seg 3 rektangler i programmet. Bredden til hvert rektangel blir

$$
h = \frac{4 - 0}{3} = \frac{4}{3}.
$$

Videre har vi at høyden til hvert rektangel er gitt ved funksjonsverdien i venstre hjørne. De tre hjørnene vi er ute etter blir

$$
x in {a, a + h, a + 2h} = {0, 4/3, 8/3}.
$$

Funksjonsverdiene i disse punktene blir

$$
f(0) = 10, \quad f(4/3) \approx 8.81, \quad f(8/3) \approx 25.41.
$$

som gir arealet

$$
A = h \cdot (f(0) + f(4/3) + f(8/3)) = \frac{4}{3} \cdot (10 + 8.81 + 25.41) \approx 58.96
$$
````

### Oppgave 2b 💻

Skriv ferdig koden til eleven slik at det regner ut en tilnærming til arealet mellom grafen til $f$ og $x$-aksen i intervallet $[0, 4]$ ved å bruke like mange rektangler som eleven opprinnelige hadde prøvd ut. <br>
Sjekk at svaret ditt stemmer med det du regnet ut i forrige oppgave.

````{admonition} Løsningsforslag
:class: dropdown

Vi kan fullføre koden inspirert av hva vi gjorde i forrige oppgave. Da får vi et program som ser slik ut:

```python
def f(x):
    return x**3 + x**2 - 4*x + 10

a = 0
b = 4
bredde = (b - a) / 3                    # Bredde med 3 rektangler

areal = 0
for i in range(3):
    høyde = f(a + i*bredde)
    areal = areal + bredde * høyde

print(f"{areal = :.2f}")                # Skriver ut arealet med 2 desimaler
```
som gir utskriften
```console
areal = 58.96
```
som betyr at arealet av de tre rektangelene er ca. 58.96.
````

### Oppgave 2c 💻

Endre programmet slik at det regner ut en *god* tilnærming til arealet under grafen. <br>
Hva blir arealet mellom grafen til $f$ og $x$-aksen i intervallet $[0, 4]$?



````{admonition} Løsningsforslag
:class: dropdown
Vi må på antall rektangler som brukes i programmet, der flere rektangler vil gi en bedre tilnærming til arealet mellom grafen til $f$ og $x$-aksen. Under vises et forslag til programmet med 1 million rektangler:

```python
def f(x):
    return x**3 + x**2 - 4*x + 10

a = 0
b = 4
n_rektangler = 1_000_000
bredde = (b - a) / n_rektangler                    

areal = 0
for i in range(n_rektangler):
    høyde = f(a + i*bredde)
    areal = areal + bredde * høyde

print(f"{areal = :.2f}")                # Skriver ut arealet med 2 desimaler
```
som gir utskriften
```
areal = 93.33
```
som betyr at arealet mellom grafen til $f$ og $x$-aksen er omtrent $93.33$. 
````